[visionOS] spatial video don't play when MediaPlayerPrivateMediaSourceAVFObjC runs in the WebContent process

jyavenard opened the pull request:
#### 70664c036656992ccabca58b08e43e5918d8abf9
<pre>
[visionOS] spatial video don&apos;t play when MediaPlayerPrivateMediaSourceAVFObjC runs in the WebContent process
<a href="https://bugs.webkit.org/show_bug.cgi?id=303301">https://bugs.webkit.org/show_bug.cgi?id=303301</a>
<a href="https://rdar.apple.com/165954788">rdar://165954788</a>

Reviewed by Gerald Squelart.

The immersive videos couldn&apos;t have played as the required metadata were lost during the WebContent-&gt;GPU serialisation.

We add support for extracting the Immersive Metadata from the AVFormatDescription.
Our SpatialVideoMetadata and VideoProjectionMetadata were each designed to only carry the information to determine if a video was spatial or immersive.
They didn&apos;t convey the entire information.
Their actual content had they contained what was needed to re-create a CMFormatDescription would have had overlapping content.
As such, we merge both of those structures into a single ImmersiveVideoMetadata.

Added API tests.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::descriptionForTrack):
* Source/WebCore/html/track/VideoTrackConfiguration.cpp:
(WebCore::VideoTrackConfiguration::setImmersiveVideoMetadata):
(WebCore::VideoTrackConfiguration::toJSON const):
(WebCore::VideoTrackConfiguration::setSpatialVideoMetadata): Deleted.
(WebCore::VideoTrackConfiguration::setVideoProjectionMetadata): Deleted.
* Source/WebCore/html/track/VideoTrackConfiguration.h:
(WebCore::VideoTrackConfiguration::immersiveVideoMetadata const):
(WebCore::VideoTrackConfiguration::spatialVideoMetadata const): Deleted.
(WebCore::VideoTrackConfiguration::videoProjectionMetadata const): Deleted.
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::relayoutDocument):
(WebCore::InspectorDOMAgent::highlightNodeList):
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):
(WebCore::InspectorDOMAgent::destroyedNodesTimerFired):
(WebCore::videoProjectionMetadataKind):
(WebCore::InspectorDOMAgent::getMediaStats):
* Source/WebCore/platform/SpatialVideoMetadata.h: Removed.
* Source/WebCore/platform/TrackInfo.h:
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModelClient::immersiveVideoMetadataChanged):
(WebCore::PlaybackSessionModelClient::spatialVideoMetadataChanged): Deleted.
(WebCore::PlaybackSessionModelClient::videoProjectionMetadataChanged): Deleted.
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::maybeUpdateVideoMetadata):
* Source/WebCore/platform/graphics/ImmersiveVideoMetadata.h: Added.
(WebCore::CameraCalibration::operator== const):
(WebCore::ImmersiveVideoMetadata::isSpatial const):
(WebCore::ImmersiveVideoMetadata::isImmersive const):
(WTF::LogArgument&lt;WebCore::ImmersiveVideoMetadata&gt;::toString):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::convertEnumerationToString):
(WebCore::convertImmersiveVideoMetadataToString):
(WebCore::convertSpatialVideoMetadataToString): Deleted.
(WebCore::convertVideoProjectionMetadataToString): Deleted.
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
* Source/WebCore/platform/graphics/VideoProjectionMetadata.h: Removed.
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::videoTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::immersiveVideoMetadata const):
(WebCore::AVTrackPrivateAVFObjCImpl::spatialVideoMetadata const): Deleted.
(WebCore::AVTrackPrivateAVFObjCImpl::videoProjectionMetadata const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
(WTF::makeVectorElement):
(WebCore::codecFromFormatDescription):
(WebCore::toHeroEye):
(WebCore::toViewPackingKind):
(WebCore::toLensAlgorithmKind):
(WebCore::toLensDomain):
(WebCore::toLensRole):
(WebCore::toExtrinsicOriginSource):
(WebCore::toCameraCalibrationDataLensCollection):
(WebCore::toVideoProjectionMetadataKind):
(WebCore::immersiveVideoMetadataFromFormatDescription):
(WebCore::extractImmersiveVideoMetadata):
(WebCore::formatDescriptionDictionaryFromImmersiveVideoMetadata):
(WebCore::spatialVideoMetadataFromFormatDescription): Deleted.
(WebCore::videoProjectionMetadataFromFormatDescription): Deleted.
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createVideoInfoFromFormatDescription):
* Source/WebCore/platform/graphics/cocoa/NullVideoPresentationInterface.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::updateConfiguration):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::cfDictionaryCreateCombined):
(WebCore::copyDescriptionExtensionValuesIfNeeded):
(WebCore::canCopyFormatDescriptionExtension): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
(WI.MediaDetailsSidebarPanel.prototype.setVideo):
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(WebKit::PlaybackSessionInterfaceLMK::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionInterfaceLMK::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionInterfaceLMK::videoProjectionMetadataChanged): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionManagerProxy::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionModelContext::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionModelContext::videoProjectionMetadataChanged): Deleted.
(WebKit::PlaybackSessionManagerProxy::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionManagerProxy::videoProjectionMetadataChanged): Deleted.
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaSpatialVideoMetadata.metadata):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionInterfaceContext::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionManager::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionInterfaceContext::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionInterfaceContext::videoProjectionMetadataChanged): Deleted.
(WebKit::PlaybackSessionManager::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionManager::videoProjectionMetadataChanged): Deleted.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ImmersiveVideoMetadata.mm: Added.
(TestWebKitAPI::compareCameraCalibrationDictionary):
(TestWebKitAPI::compareCameraCalibrationArray):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, Spatial)):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, ImmersiveMono)):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, ImmersiveStereo_SideBySide)):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, ImmersiveStereo_OverUnder)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/immersivemono.mp4: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/immersivestereo-overunder.mp4: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/immersivestereo-sbs.mp4: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/spatial.mp4: Added.

Canonical link: <a href="https://commits.webkit.org/304503@main">https://commits.webkit.org/304503@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bafc8fedab56f53e5b6432054e23f6c7cd0dc4b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135774 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8150 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47067 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143488 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/87399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/de9e0675-7786-439a-b559-cb1ae91243df) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137643 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8808 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7997 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103762 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/87399 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138720 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6350 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121696 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84640 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/6097 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3717 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4094 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127756 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115327 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39885 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146237 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/134257 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7838 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40452 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112123 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7864 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6577 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112504 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28543 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5976 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117997 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/61768 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7884 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36102 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167078 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7618 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71434 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43616 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7845 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7713 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/UserInterface/Protocol/Legacy/macOS/26.0/InspectorBackendCommands.js:
```diff
@@ -178,7 +178,7 @@ InspectorBackend.registerEnum("DOM.AccessibilityPropertiesCurrent", {True: "true
 InspectorBackend.registerEnum("DOM.AccessibilityPropertiesLiveRegionStatus", {Assertive: "assertive", Polite: "polite", Off: "off"});
 InspectorBackend.registerEnum("DOM.AccessibilityPropertiesSwitchState", {Off: "off", On: "on"});
-InspectorBackend.registerEnum("DOM.VideoProjectionMetadataKind", {Unknown: "unknown", Equirectangular: "equirectangular", HalfEquirectangular: "half-equirectangular", EquiAngularCubemap: "equi-angular-cubemap", Parametric: "parametric", Pyramid: "pyramid", AppleImmersiveVideo: "apple-immersive-video"});
+InspectorBackend.registerEnum("DOM.VideoProjectionMetadataKind", {Unknown: "unknown", Rectilinear: "rectilinear", Equirectangular: "equirectangular", HalfEquirectangular: "half-equirectangular", EquiAngularCubemap: "equi-angular-cubemap", Parametric: "parametric", Pyramid: "pyramid", AppleImmersiveVideo: "apple-immersive-video"});
```
please do not modify any files in `Source/WebInspectorUI/Protocol` as they are a snapshot of what was supported in that version of WebKit (i.e. these files are how Web Inspector is able to support inspecting an older version of iOS that lack newer features)

jyavenard replied:
Adding a value for an enum, shouldn't have any negative impact right ?

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
```diff
@@ -343,9 +344,9 @@ WI.MediaDetailsSidebarPanel = class MediaDetailsSidebarPanel extends WI.DOMDetai
-        this.#ui.fovRow.pendingValue = WI.UIString("%dº").format(video?.spatialVideoMetadata?.horizontalFOVDegrees ?? 0);
-        this.#ui.baselineRow.pendingValue = WI.UIString("%dmm").format((video?.spatialVideoMetadata?.baseline ?? 0) / 1000);
-        this.#ui.disparityRow.pendingValue = WI.UIString("%d%%").format((video?.spatialVideoMetadata?.disparityAdjustment ?? 0) * 100);
+        this.#ui.fovRow.pendingValue = WI.UIString("%dº").format((video?.spatialVideoMetadata?.horizontalFieldOfView ?? 0) / 1000);
```
have these properties been shipped to an iOS release yet? if so, we'll need to keep these old properties for that like
```js
video?.spatialVideoMetadata?.horizontalFieldOfView ?? video?.spatialVideoMetadata?.horizontalFOVDegrees
```

squelart commented on Source/WebCore/platform/TrackInfo.h:
```diff
@@ -142,6 +143,10 @@ struct VideoInfo : public TrackInfo {
     PlatformVideoColorSpace colorSpace;
 
+#if PLATFORM(VISION)
+    std::optional<ImmersiveVideoMetadata> immersiveVideoMetadata;
```
ImmersiveVideoMetadata is not small, so are you okay to always take this much space (plus the optional's flag) here and other places?
Alternatively, it could be stored away through a unique_ptr, but manipulation would of course be more "interesting". 😄

jyavenard replied:
TrackInfo/VideoInfo are created on the heap.A VideoInfo is refcounted. It’s reused across all MediaSample. Be it on the stack or on the heap makes no difference here, except that we would have two call to mallocs instead of just one.

squelart commented on Source/WebCore/platform/graphics/ImmersiveVideoMetadata.h:
```diff
@@ -0,0 +1,165 @@
+    StereoCameraSystemBaseline,
+};
+
+using RadialDistortionCoefficients = Vector<float, 4>;
+using LensFrameAdjustmentsPolynomial = Vector<float, 3>;
+using ExtrinsicOrientationQuaternion = Vector<float, 3>;
```
Are these really `Vector`s, as in resizable arrays?
Otherwise if they're always fixed-size, consider using std::array instead.

jyavenard replied:
They are supposed to be fixed size. However they are constructed from a CFArray which is dynamic in size. So to ensure we don’t get OOB I made them Vector instead.

squelart commented on Source/WebCore/platform/graphics/MediaSourcePrivate.cpp:
```diff
@@ -192,7 +192,8 @@ void MediaSourcePrivate::updateTracksType()
     TracksType tracksCombinedTypes;
     for (auto type : m_tracksTypes.values())
         tracksCombinedTypes |= type;
-    m_tracksCombinedTypes = tracksCombinedTypes;
+    if (m_tracksCombinedTypes.exchange(tracksCombinedTypes) != tracksCombinedTypes) {
+    }
```
What's the point of this `if` with an empty block?

jyavenard replied:
It’s a rebase error of another PR I just landed.

squelart commented on Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
```diff
@@ -245,96 +252,397 @@ bool formatDescriptionIsProtected(CMFormatDescriptionRef formatDescription)
-    if (PAL::canLoad_CoreMedia_kCMFormatDescriptionProjectionKind_AppleImmersiveVideo()
-        && CFStringCompare(kind, PAL::kCMFormatDescriptionProjectionKind_AppleImmersiveVideo, 0) == kCFCompareEqualTo)
-        return VideoProjectionMetadata::Kind::AppleImmersiveVideo;
+    if (CFStringCompare(kind, PAL::kCMFormatDescriptionProjectionKind_ParametricImmersive, 0) == kCFCompareEqualTo)
+        return ImmersiveVideoMetadata::Kind::Parametric;
```
So kCMFormatDescriptionProjectionKind_ParametricImmersive corresponds ImmersiveVideoMetadata::Kind::Parametric, is the shorter name without "Immersive" intended? (Other ones seem to match fully.)

jyavenard replied:
Yes. Parametric is parametric immersive.

jyavenard replied:
I'm just re-using the existing enum here.

squelart reviewed (commented):
One strange `if`, and minor nits.

squelart commented on Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
```diff
@@ -113,4 +113,42 @@ void ArgumentCoder<StringView>::encode<Encoder>(Encoder&, StringView);
+    auto col2row0 = decoder.template decode<float>();
+    auto col2row1 = decoder.template decode<float>();
+    auto col2row2 = decoder.template decode<float>();
+    if (!col0row0 || !col0row1 || !col0row2 || !col1row0 || !col1row1 || !col1row2 || !col2row0 || !col2row1 || !col2row2) [[unlikely]]
```
Is it safe to call decode multiple times without checking in-between?
After a quick search, I don't see any other code like this.

It should be possible to turn this into a loop like in `encode()` above, don't you think?

jyavenard replied:
Yes it's fine. The décoder will be marked as failed. There's no need to test each time.

squelart reviewed (approved):
Thank you.
One final nit, please either check every decode (unrolled or in a loop), or prove that it's not necessary.

sheeparegreat commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -5695,17 +5695,69 @@ enum class WebCore::PlatformVideoMatrixCoefficients : uint8_t {
+using WebCore::LensFrameAdjustmentsPolynomial = Vector<float, 3>;
+using WebCore::ExtrinsicOrientationQuaternion = Vector<float, 3>;
+#if PLATFORM(COCOA)
+using WebCore::IntrinsicMatrix = matrix_float3x3;
```
`matrix_float3x3` should use generated serialization.

sheeparegreat commented on Source/WebKit/Platform/IPC/ArgumentCoders.cpp:
```diff
@@ -113,4 +113,42 @@ void ArgumentCoder<StringView>::encode<Encoder>(Encoder&, StringView);
 
+#if __has_include(<simd/simd.h>)
+template<typename Encoder>
+void ArgumentCoder<matrix_float3x3>::encode(Encoder& encoder, const matrix_float3x3& matrix)
```
Serialization for this type should be generated from the IDL too. We should not have any manually specified serializers.

sheeparegreat reviewed (changes requested):
We should not have any manual encoders, all serializers should be generated from IDL.   I see in the PR we have a new manual encoder created for `matrix_float3x3>`.  Please describe this type in the IDL such that the serializers are generated.

sheeparegreat reviewed (commented):
LGTM from the CoreIPC point of view. Thank you for the update.

jyavenard committed:
[visionOS] spatial video don't play when MediaPlayerPrivateMediaSourceAVFObjC runs in the WebContent process
https://bugs.webkit.org/show_bug.cgi?id=303301
rdar://165954788

Reviewed by Gerald Squelart.

The immersive videos couldn't have played as the required metadata were lost during the WebContent->GPU serialisation.

We add support for extracting the Immersive Metadata from the AVFormatDescription.
Our SpatialVideoMetadata and VideoProjectionMetadata were each designed to only carry the information to determine if a video was spatial or immersive.
They didn't convey the entire information.
Their actual content had they contained what was needed to re-create a CMFormatDescription would have had overlapping content.
As such, we merge both of those structures into a single ImmersiveVideoMetadata.

Added API tests.

* Source/JavaScriptCore/inspector/protocol/DOM.json:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.cpp:
* Source/WebCore/PAL/pal/cf/CoreMediaSoftLink.h:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::descriptionForTrack):
* Source/WebCore/html/track/VideoTrackConfiguration.cpp:
(WebCore::VideoTrackConfiguration::setImmersiveVideoMetadata):
(WebCore::VideoTrackConfiguration::toJSON const):
(WebCore::VideoTrackConfiguration::setSpatialVideoMetadata): Deleted.
(WebCore::VideoTrackConfiguration::setVideoProjectionMetadata): Deleted.
* Source/WebCore/html/track/VideoTrackConfiguration.h:
(WebCore::VideoTrackConfiguration::immersiveVideoMetadata const):
(WebCore::VideoTrackConfiguration::spatialVideoMetadata const): Deleted.
(WebCore::VideoTrackConfiguration::videoProjectionMetadata const): Deleted.
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::relayoutDocument):
(WebCore::InspectorDOMAgent::highlightNodeList):
(WebCore::InspectorDOMAgent::buildObjectForAccessibilityProperties):
(WebCore::InspectorDOMAgent::destroyedNodesTimerFired):
(WebCore::videoProjectionMetadataKind):
(WebCore::InspectorDOMAgent::getMediaStats):
* Source/WebCore/platform/SpatialVideoMetadata.h: Removed.
* Source/WebCore/platform/TrackInfo.h:
(WebCore::VideoInfo::create):
(WebCore::VideoInfo::VideoInfo):
* Source/WebCore/platform/cocoa/PlaybackSessionModel.h:
(WebCore::PlaybackSessionModelClient::immersiveVideoMetadataChanged):
(WebCore::PlaybackSessionModelClient::spatialVideoMetadataChanged): Deleted.
(WebCore::PlaybackSessionModelClient::videoProjectionMetadataChanged): Deleted.
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.h:
* Source/WebCore/platform/cocoa/PlaybackSessionModelMediaElement.mm:
(WebCore::PlaybackSessionModelMediaElement::maybeUpdateVideoMetadata):
* Source/WebCore/platform/graphics/ImmersiveVideoMetadata.h: Added.
(WebCore::CameraCalibration::operator== const):
(WebCore::ImmersiveVideoMetadata::isSpatial const):
(WebCore::ImmersiveVideoMetadata::isImmersive const):
(WTF::LogArgument<WebCore::ImmersiveVideoMetadata>::toString):
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::convertEnumerationToString):
(WebCore::convertImmersiveVideoMetadataToString):
(WebCore::convertSpatialVideoMetadataToString): Deleted.
(WebCore::convertVideoProjectionMetadataToString): Deleted.
* Source/WebCore/platform/graphics/PlatformVideoTrackConfiguration.h:
* Source/WebCore/platform/graphics/VideoProjectionMetadata.h: Removed.
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.h:
* Source/WebCore/platform/graphics/avfoundation/AVTrackPrivateAVFObjCImpl.mm:
(WebCore::AVTrackPrivateAVFObjCImpl::videoTrackConfiguration const):
(WebCore::AVTrackPrivateAVFObjCImpl::immersiveVideoMetadata const):
(WebCore::AVTrackPrivateAVFObjCImpl::spatialVideoMetadata const): Deleted.
(WebCore::AVTrackPrivateAVFObjCImpl::videoProjectionMetadata const): Deleted.
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.cpp:
(WTF::makeVectorElement):
(WebCore::codecFromFormatDescription):
(WebCore::toHeroEye):
(WebCore::toViewPackingKind):
(WebCore::toLensAlgorithmKind):
(WebCore::toLensDomain):
(WebCore::toLensRole):
(WebCore::toExtrinsicOriginSource):
(WebCore::toCameraCalibrationDataLensCollection):
(WebCore::toVideoProjectionMetadataKind):
(WebCore::immersiveVideoMetadataFromFormatDescription):
(WebCore::extractImmersiveVideoMetadata):
(WebCore::formatDescriptionDictionaryFromImmersiveVideoMetadata):
(WebCore::spatialVideoMetadataFromFormatDescription): Deleted.
(WebCore::videoProjectionMetadataFromFormatDescription): Deleted.
* Source/WebCore/platform/graphics/avfoundation/FormatDescriptionUtilities.h:
* Source/WebCore/platform/graphics/cocoa/CMUtilities.mm:
(WebCore::createFormatDescriptionFromTrackInfo):
(WebCore::createVideoInfoFromFormatDescription):
* Source/WebCore/platform/graphics/cocoa/NullVideoPresentationInterface.h:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
* Source/WebCore/platform/graphics/cocoa/VideoTrackPrivateWebM.cpp:
(WebCore::VideoTrackPrivateWebM::updateConfiguration):
* Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
(WebCore::cfDictionaryCreateCombined):
(WebCore::copyDescriptionExtensionValuesIfNeeded):
(WebCore::canCopyFormatDescriptionExtension): Deleted.
* Source/WebCore/platform/ios/VideoPresentationInterfaceIOS.h:
* Source/WebInspectorUI/UserInterface/Views/MediaDetailsSidebarPanel.js:
(WI.MediaDetailsSidebarPanel.prototype.setVideo):
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.h:
* Source/WebKit/Platform/ios/PlaybackSessionInterfaceLMK.mm:
(WebKit::PlaybackSessionInterfaceLMK::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionInterfaceLMK::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionInterfaceLMK::videoProjectionMetadataChanged): Deleted.
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/PlaybackSessionManagerProxy.mm:
(WebKit::PlaybackSessionModelContext::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionManagerProxy::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionModelContext::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionModelContext::videoProjectionMetadataChanged): Deleted.
(WebKit::PlaybackSessionManagerProxy::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionManagerProxy::videoProjectionMetadataChanged): Deleted.
* Source/WebKit/UIProcess/Cocoa/VideoPresentationManagerProxy.h:
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController configureEnvironmentPickerOrFullscreenVideoButtonView]):
* Source/WebKit/WebKitSwift/LinearMediaKit/LinearMediaTypes.swift:
(WKSLinearMediaSpatialVideoMetadata.metadata):
* Source/WebKit/WebKitSwift/LinearMediaKit/WKSLinearMediaTypes.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.h:
* Source/WebKit/WebProcess/cocoa/PlaybackSessionManager.mm:
(WebKit::PlaybackSessionInterfaceContext::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionManager::immersiveVideoMetadataChanged):
(WebKit::PlaybackSessionInterfaceContext::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionInterfaceContext::videoProjectionMetadataChanged): Deleted.
(WebKit::PlaybackSessionManager::spatialVideoMetadataChanged): Deleted.
(WebKit::PlaybackSessionManager::videoProjectionMetadataChanged): Deleted.
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ImmersiveVideoMetadata.mm: Added.
(TestWebKitAPI::compareCameraCalibrationDictionary):
(TestWebKitAPI::compareCameraCalibrationArray):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, Spatial)):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, ImmersiveMono)):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, ImmersiveStereo_SideBySide)):
(TestWebKitAPI::TEST(ImmersiveVideoMetadata, ImmersiveStereo_OverUnder)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/immersivemono.mp4: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/immersivestereo-overunder.mp4: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/immersivestereo-sbs.mp4: Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/spatial.mp4: Added.

Canonical link: https://commits.webkit.org/304503@main
